File: backend/app/src/main/kotlin/net/animeta/backend/controller/v2/TablePeriodController.kt
package net.animeta.backend.controller.v2

import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder
import net.animeta.backend.db.Datastore
import net.animeta.backend.db.query
import net.animeta.backend.dto.WorkDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.Period
import net.animeta.backend.model.QWorkPeriodIndex.workPeriodIndex
import net.animeta.backend.model.User
import net.animeta.backend.repository.RecordRepository
import net.animeta.backend.repository.UserRepository
import net.animeta.backend.security.CurrentUser
import net.animeta.backend.serializer.RecordSerializer
import net.animeta.backend.serializer.WorkSerializer
import net.animeta.backend.service.RecommendationService
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.time.ZoneId
import java.util.concurrent.TimeUnit

@RestController
@RequestMapping("/v2/table/periods/{period:[0-9]{4}Q[1-4]}")
class TablePeriodController(val datastore: Datastore,
                            val userRepository: UserRepository,
                            val recordRepository: RecordRepository,
                            val workSerializer: WorkSerializer,
                            val recordSerializer: RecordSerializer,
                            val recommendationService: RecommendationService) {
    private val cache: Cache<Period, List<WorkDTO>> = CacheBuilder.newBuilder()
            .expireAfterWrite(1, TimeUnit.HOURS)
            .build()
    private val defaultTimeZone = ZoneId.of("Asia/Seoul")
    private val minPeriod = Period(2014, 2)

    fun invalidateCache() {
        cache.invalidateAll()
        recommendationService.invalidateCache()
    }

    @GetMapping
    fun get(@PathVariable("period") periodParam: String,
            @RequestParam("with_recommendations", defaultValue = "false") withRecommendations: Boolean,
            @RequestParam("only_added", defaultValue = "false") onlyAdded: Boolean,
            @RequestParam("username", required = false) username: String?,
            @CurrentUser(required = false) currentUser: User?): List<WorkDTO> {
        val period = Period.parse(periodParam) ?: throw ApiException.notFound()
        val maxPeriod = Period.now(defaultTimeZone).next()
        if (period < minPeriod || period > maxPeriod) {
            throw ApiException.notFound()
        }
        var result = cache.get(period) {
            var query = workPeriodIndex.query
                    .filter(workPeriodIndex.period.eq(period.toString()))
            query = query.filter(workPeriodIndex.firstPeriod.isTrue)
            datastore.query(query).map { workSerializer.serialize(it.work) }
        }
        val user = username?.let { userRepository.findByUsername(it) } ?: currentUser
        if (user == null) {
            return result
        }
        val records = recordRepository.findAllByUserAndWorkIdIn(user, result.map { it.id })
                .map { recordSerializer.serialize(it, RecordSerializer.legacyOptions()) }
                .associateBy { it.work_id }
        val recommendationContext = if (withRecommendations && currentUser != null) {
            recommendationService.createContext(result.map { it.id }, currentUser)
        } else {
            null
        }
        if (onlyAdded) {
            result = result.filter { it.id in records }
        }
        return result.map {
            val record = records[it.id]
            if (recommendationContext != null) {
                val (recommendations, recommendationScore) = recommendationService.generate(it.id, recommendationContext)
                it.copy(
                    record = record,
                    recommendations = recommendations,
                    recommendationScore = recommendationScore
                )
            } else {
                it.copy(record = record)
            }
        }
    }
}

File: backend-ts/src/controllers/table_period.controller.ts
import { CACHE_MANAGER, Controller, DefaultValuePipe, Get, Inject, Param, ParseBoolPipe, Query } from "@nestjs/common";
import { CurrentUser } from "src/auth/decorators";
import { User } from "src/entities/user.entity";
import { ApiException } from "./exceptions";
import { WorkSerializer } from "src/serializers/work.serializer";
import { RecordDTO, WorkDTO } from "shared/types";
import { Period } from "src/utils/period";
import { Temporal } from "@js-temporal/polyfill";
import { Cache } from "cache-manager";
import { Repository } from "typeorm";
import { Work } from "src/entities/work.entity";
import { InjectRepository } from "@nestjs/typeorm";
import { RecordService } from "src/services/record.service";
import { RecordSerializer } from "src/serializers/record.serializer";
import { RecommendationService } from "src/services/recommendation.service";

@Controller('/api/v4/table/periods/:period([0-9]{4}Q[1-4])')
export class TablePeriodController {
  private defaultTimeZone = Temporal.TimeZone.from('Asia/Seoul')
  private minPeriod = new Period(2014, 2)

  constructor(
    @InjectRepository(Work) private workRepository: Repository<Work>,
    @InjectRepository(User) private userRepository: Repository<User>,
    private recordService: RecordService,
    private recommendationService: RecommendationService,
    private workSerializer: WorkSerializer,
    private recordSerializer: RecordSerializer,
    @Inject(CACHE_MANAGER) private cache: Cache,
  ) {}

  @Get()
  async get(
    @Param('period') periodParam: string,
    @Query('with_recommendations', new DefaultValuePipe('false'), new ParseBoolPipe()) withRecommendations: boolean,
    @Query('only_added', new DefaultValuePipe('false'), new ParseBoolPipe()) onlyAdded: boolean, 
    @Query('username') username: string | null,
    @CurrentUser({ required: false }) currentUser: User | null,
  ): Promise<WorkDTO[]> {
    const period = Period.parse(periodParam)
    if (!period) throw ApiException.notFound()
    const maxPeriod = Period.now(this.defaultTimeZone).next()
    if (period.compareTo(this.minPeriod) < 0 || period.compareTo(maxPeriod) > 0) {
      throw ApiException.notFound()
    }
    const cacheKey = `table:${period}`
    let result = await this.cache.wrap<WorkDTO[]>(cacheKey, async () => {
      const works = await this.workRepository.find({ where: {first_period: period.toString()} })
      return Promise.all(works.map(work => this.workSerializer.serialize(work)))
    }, { ttl: 60 * 60 })
    const user = username ? await this.userRepository.findOne({ where: {username} }) : currentUser
    if (!user) return result
    const records = await Promise.all(result.map(async it => {
      const record = await this.recordService.findByUserAndWorkId(user, it.id)
      if (!record) return null
      return await this.recordSerializer.serialize(record, RecordSerializer.legacyOptions())
    }))
    const recordsByWorkId = new Map<number, RecordDTO>(records.filter(it => it).map(it => [it!.work_id, it!]))
    const recommendationContext = withRecommendations && user ? await this.recommendationService.createContext(result.map(it => it.id), user) : null
    if (onlyAdded) {
      result = result.filter(it => recordsByWorkId.has(it.id))
    }
    return Promise.all(result.map(async it => {
      const record = recordsByWorkId.get(it.id) ?? null
      if (recommendationContext) {
        const { recommendations, recommendationScore } = await this.recommendationService.generate(it.id, recommendationContext)
        return {...it, record, recommendations, recommendationScore}
      } else {
        return {...it, record}
      }
    }))
  }
}
